core/state/workers.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import "runtime"
+
+// workers is a singleton pool of goroutines to run arbitrary tasks concurrently.
+var workers = newWorkerPool(16 * runtime.GOMAXPROCS(0))
```

Comment:
Why `16 x ` the number of CPUs? Why not just `newWorkerPool(runtime.GOMAXPROCS(0))` ?

Reply:
Ah, so the pool still only uses `runtime.GOMAXPROCS` channels, but the queue has some slack so we don't have to wait for a thread to be ready.

Reply:
Yeah, that was the rationale. Another alternative could be to make the pool even fancier and support a Join operation. So each "source" can queue up tasks (maybe unlimited), but then they could also join the pool in crunching through them (only own tasks) to ensure that if the pool is busy on another task (e.g. prefetcher, rpc?, tracer), then the main thread doesn't wait, rather progresses guaranteed. Not sure, only ideas here.